## PostgreSQL
### Exercices

In [ ]:
import pandas as pd

df = pd.read_csv("/Users/polo11/Downloads/bank_full.csv")
df.columns

In [38]:
import psycopg2

# Connect to PostgreSQL database
conn = psycopg2.connect(dbname="new_db", user="postgres", password="PF1234-5678", host="localhost",  port="5432" )
c = conn.cursor()

In [39]:
##1: What is the average age of the customers?
c.execute("""SELECT AVG(age) FROM bank_full""")
average_age = c.fetchone()[0]
average_age_phrase = "L'âge moyen est de {:.2f} ans.".format(average_age)
print(average_age_phrase)

L'âge moyen est de 40.94 ans.


In [40]:
##2: How many customers have defaulted on loans?
conn = psycopg2.connect(dbname="new_db", user="postgres", password="PF1234-5678", host="localhost",  port="5432" )
c = conn.cursor()
c.execute(""" SELECT COUNT(*) 
          FROM bank_full
          WHERE default_='yes' """)
c.fetchall()


[(815,)]

Il y a **815** clients en défaut de paiement.

In [41]:
##3: What is the maximum balance in the account?
c.execute(""" SELECT MAX(balance) FROM bank_full""")
c.fetchall()

[(Decimal('102127'),)]

La balance maximale est de **102 127€**.

In [42]:
##4: How many customers have a housing loan?
c.execute(""" SELECT COUNT(*) FROM bank_full
          WHERE housing = 'yes' """)
c.fetchall()

[(25130,)]

**25 130** clients ont un prêt immobilier.

In [43]:
##5: What is the education level of the youngest customer?
c.execute(""" SELECT education, MIN(age)
          FROM bank_full
          GROUP BY education
          LIMIT 1""")
c.fetchall()

[('secondary', 18)]

Le niveau d'éducation du plus jeune client est un niveau **secondaire**.

In [44]:
##6: What is the marital status of the oldest customer?
c.execute(""" SELECT MAX(age), marital
          FROM bank_full
          GROUP BY marital
          ORDER BY MAX(age) DESC
          LIMIT 1""")
c.fetchall()

[(95, 'divorced')]

Le client le plus âgé a **95 ans** et est **divorcé**.

In [45]:
##7: What is the distribution of the months in which customers were last contacted?
c.execute(""" SELECT COUNT(*) as num_customers, month
          FROM bank_full
          WHERE month IS NOT NULL
          GROUP BY month
          ORDER BY num_customers DESC""")
c.fetchall()

[(13766, 'may'),
 (6895, 'jul'),
 (6247, 'aug'),
 (5341, 'jun'),
 (3970, 'nov'),
 (2932, 'apr'),
 (2649, 'feb'),
 (1403, 'jan'),
 (738, 'oct'),
 (579, 'sep'),
 (477, 'mar'),
 (214, 'dec')]

In [46]:
##8: What is the average duration of contact for customers who have a personal loan?
c.execute(""" SELECT AVG(duration) as avg_duration FROM bank_full
          WHERE loan='yes' """)
c.fetchall()

[(Decimal('250.8453892876863611'),)]

La durée moyenne de contact pour les gens qui ont un prêt perosnnel est de **250.85 jours**.

In [47]:
##9: How many customers were previously contacted, and what was the average duration of their last contact?
c.execute(""" SELECT COUNT(*) as num_customers, AVG(duration) as avg_duration FROM bank_full
          WHERE pdays != -1 """)
c.fetchall()

[(8257, Decimal('260.1186871745185903'))]

Il y a **8257** clients qui ont été contactés et la durée moyenne de leur dernier contact est de **260 jours**.

In [48]:
##10: What is the distribution of job types among customers?
c.execute(""" SELECT job , COUNT(*) as num_customers FROM bank_full
          GROUP BY job
          ORDER BY num_customers DESC""")
c.fetchall()

[('blue-collar', 9732),
 ('management', 9458),
 ('technician', 7597),
 ('admin.', 5171),
 ('services', 4154),
 ('retired', 2264),
 ('self-employed', 1579),
 ('entrepreneur', 1487),
 ('unemployed', 1303),
 ('housemaid', 1240),
 ('student', 938),
 ('unknown', 288)]

In [49]:
##11: How many customers have been contacted more than once in the current campaign?
c.execute(""" SELECT COUNT(*) as num_customers FROM bank_full
          WHERE campaign >1 """)
c.fetchall()

[(27667,)]

Il y a **27667** clients qui ont été contacté plus d'une fois dans la campagne actuelle.

In [50]:
##12: What is the success rate of previous marketing campaigns (poutcome) among customers.
c.execute(""" SELECT poutcome, COUNT(*) as num_customers
          FROM bank_full
          GROUP BY poutcome""")
result1 = c.fetchall()

c.execute(""" SELECT COUNT(*) as success
          FROM bank_full
          WHERE poutcome='success'""")
result2 = c.fetchall()

#On calcul le taux de réussite des campagnes marketing précédentes.
print(result2[0][0]/result1[0][1])

0.821195652173913


Le taux de réussite des campagnes marketing précédentes sur les clientse est de **82.12%**.